### ISSUE TITLE
assertSame() specifies a value

### ISSUE THREAD

**Issue by mhujer** [phpstan/phpstan-phpunit#17]
I just tried `dev-master` version of PHPStan with `dev-master` (actually `^0.10`) of phpstan-phpunit and I stumbled across this issue: assertSame somehow fixes the value to the first occurrence.

Although the test pass, an error is reported for the second assert:
```
Call to static method PHPUnit\Framework\Assert::assertSame() with int(200) and int(100) will always evaluate to false.
```

```php
<?php declare(strict_types = 1);

namespace App;

class FooTest extends \PHPUnit\Framework\TestCase
{

	public function testAssertSameBehaviour(): void
	{
		$foo = new Foo();
		$foo->setValue(100);

		self::assertSame(100, $foo->getValue());

		$foo->setValue(200);
		self::assertSame(200, $foo->getValue());
	}

}


class Foo {

	/** @var int */
	private $value;

	public function getValue(): int
	{
		return $this->value;
	}

	public function setValue(int $value): void
	{
		$this->value = $value;
	}

}
```

**Comment by ondrejmirtes** [comment]
Hi, nice catch! I didn't think of this use case and I'm not sure what to do about it. Actually, it has nothing to do with `phpstan/phpstan-phpunit`, it's just a "feature" of TypeSpecifier combined with constant types, see: https://phpstan.org/r/882c0501428e1146c68172259801a5b2

**Comment by mhujer** [comment]
And also "NEVER" can happen: `Strict comparison using === between *NEVER* and int(300) will always evaluate to false. `: https://phpstan.org/r/aa260e1dbd0e7839e8f2c259d845b51c

**Comment by ondrejmirtes** [comment]
Makes sense, since for PHPStan, it's guaranteed that `getValue()` will return a specific value after each `assert`.

But this behaviour is obviously wrong. The main problem is with specifying what a method will return, and relying on that it will do that next time too. I could turn off this behaviour but that would break a lot of things since a lot of code relies on this and in different cases, this works.

The correct solution would be some deep and recursive introspection of the code but that's outside of the scope of the next version.

Another possible solution is to specify only type and not a concrete value, since this problem became more apparent with constant values in 0.10-dev:

```php
assert($foo->getValue() === 100);
// $foo->getValue() is definitely int, not int(100)
```

But I can easily imagine code that will be broken too:

```php
// getValue(): ?int
assert($foo->getValue() === 100); // $foo->getValue() is definitely int, eliminated null
$foo->setValue(null);
assert($foo->getValue() === null); // "will always evaluate to false"
```

What can be done in the scope of PHPStan 0.10, but requires more work from users, is to write another type-specifying extension that will specify "after calling `setValue()`, here's what `getValue()` will return". Of course, there could be a general solution that would just require putting your classes with setters and getters into a config, similarly  to `universalObjecrCrateClasses`, but it's still more work. But it would keep everything consistent and it would work.

There would be a minor problem with that - with `checkAlwaysTrueCheckTypeFunctionCall: true` (which is the default when you have `strict-rules` installed), asserting the obvious will report:

```php
$foo = new Foo();
$foo->setValue(100);
self::assertSame(100, $foo->getValue()); // "will always evaluate to true"
```

But that can be turned off and my argument is "don't test the obvious" :)

So, what do you think?

**Comment by mhujer** [comment]
I actually discovered the issue in a controller test (which uses Symfony `WebTestCase`) and the example above was only the simplest case I was able to reproduce the issue in.

 ```php
<?php
class RegistrationTest extends \Symfony\Bundle\FrameworkBundle\Test\WebTestCase
{

	public function testCanBeRegistered(): void
	{
		$client = static::createClient();

		$crawler = $client->request('GET', '/some/page/');
		self::assertSame(200, $client->getResponse()->getStatusCode());

		$sendButton = $crawler->selectButton('registration_form[register]');
		$registrationForm = $sendButton->form();

		$client->submit($registrationForm, [
			'registration_form[name]' => 'Name',
			'registration_form[email]' => 'some-email@gmail.com',
		]);

		// Call to static method PHPUnit\Framework\Assert::assertSame() with int(302) and int(200) will always evaluate to false.
		self::assertSame(302, $client->getResponse()->getStatusCode());
		$crawler = $client->followRedirect();

		//  Call to static method PHPUnit\Framework\Assert::assertSame() with int(200) and *NEVER* will always evaluate to false.
		self::assertSame(200, $client->getResponse()->getStatusCode());
		
		// ....
	}
}
```

I'm afraid that it is not possible statically analyze what the response code would be in this case.

**Comment by ondrejmirtes** [comment]
Tried to solve it with https://github.com/phpstan/phpstan/commit/4d22156d4fc8c2b5a04c1c63c21cb42d69b22686, please test it 😊 Thanks!

**Comment by mhujer** [comment]
This example https://github.com/phpstan/phpstan-phpunit/issues/17#issuecomment-379523808 still reports:

```
  24     Call to static method PHPUnit\Framework\Assert::assertSame() with 302 and 200 will always evaluate to false.
  28     Call to static method PHPUnit\Framework\Assert::assertSame() with 200 and *NEVER* will always evaluate to false.
```

I have latest dev versions installed:

```
λ composer show phpstan/*

phpstan/phpdoc-parser        0.2                PHPDoc parser with support for nullable, intersection and generic types
phpstan/phpstan              dev-master 2ff7156 PHPStan - PHP Static Analysis Tool
phpstan/phpstan-phpunit      dev-master a0354f8 PHPUnit extensions and rules for PHPStan
phpstan/phpstan-strict-rules dev-master 2dfbf80 Extra strict and opinionated rules for PHPStan
```

**Comment by ondrejmirtes** [comment]
OK, try it now please 😊 https://github.com/phpstan/phpstan/commit/d35ef354365a186770c0ad627a93dd4a538c8f46

**Comment by mhujer** [comment]
It works fine now! 👍  Awesome 👏
